# Pythonic Data Cleaning With NumPy and Pandas

## What we are covering

- Dropping Columns in a DataFrame
- Changing the Index of a DataFrame
- Tidying up Fields in the Data
- Combining str Methods with NumPy to Clean Columns
- Cleaning the Entire Dataset Using the applymap Function
- Renaming Columns and Skipping Rows
- Python Data Cleaning: Recap and Resources



Analyst spend a large amount of their time cleaning datasets and getting them down to a form with which they can work. In fact, a lot of data scientists argue that the initial steps of obtaining and cleaning data constitute 80% of the job. So, it is important to be able to deal with messy data, whether that means missing values, inconsistent formatting, malformed records, or nonsensical outliers.

**Here are the datasets that will be used**:
 
- [BL-Flickr_Images-Book.csv](https://github.com/realpython/python-data-cleaning/blob/master/Datasets/BL-Flickr-Images-Book.csv) -  A CSV file containing information about books from the British Library.
- [University Townx text file](https://github.com/realpython/python-data-cleaning/blob/master/Datasets/university_towns.txt) - A text file containing names of college towns in every US state.
- [Olymics](https://github.com/realpython/python-data-cleaning/blob/master/Datasets/olympics.csv) - A CSV file summarizing the participation of all countries in the Summer and Winter Olypics.

Let’s import the required modules and get started!
```python
import pandas as pd
import numpy as np
```

In [1]:
import pandas as pd
import numpy as np
import os

### Dropping Columns in a DataFrame

Often, you’ll find that not all the categories of data in a dataset are useful to you. For example, you might have a dataset containing student information (name, grade, standard, parents’ names, and address) but want to focus on analyzing student grades. 

In this case, the address or parents’ names categories are not important to you. Retaining these unneeded categories will take up unnecessary space and potentially also bog down runtime.

Pandas provides a handy way of removing unwanted columns or rows from a DataFrame with the *drop()* function. Let’s look at a simple example where we drop a number of columns from a DataFrame.

#### BL-Flickr-Images-Book
First, let’s create a DataFrame out of the CSV file ‘BL-Flickr-Images-Book.csv’. In the examples below, we pass a relative path to pd.read_csv, meaning that all of the datasets are in a folder named Datasets in our current working directory:

In [2]:
# Read the csv file
df = pd.read_csv(os.path.join(os.path.abspath('.\Files'),'BL-Flickr-Images-Book.csv' ))

# print the first the head; which are the first 5
df.head()

# can you print more than 5 records using head

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [3]:
df.head(10)

Identifier              Edition Statement  \
0         206                            NaN   
1         216                            NaN   
2         218                            NaN   
3         472                            NaN   
4         480   A new edition, revised, etc.   
5         481  Fourth edition, revised, etc.   
6         519                            NaN   
7         667                            NaN   
8         874                            NaN   
9        1143                            NaN   

                  Place of Publication Date of Publication  \
0                               London         1879 [1878]   
1             London; Virtue & Yorston                1868   
2                               London                1869   
3                               London                1851   
4                               London                1857   
5                               London                1875   
6                               London                1872   
7  pp. 40. G. Bryan & Co: Oxford, 1898                 NaN   
8                              London]                1676   
9                               London                1679   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   
5      William Macintosh  [The World in which I live, and my place in it...   
6             The Author  Lagonells. By the author of Darmayne (F. E. A....   
7                    NaN  The Coming of Spring, and other poems. By J. A...   
8                    NaN  A Warning to the inhabitants of England, and L...   
9                    NaN  A Satyr against Vertue. (A poem: supposed to b...   

          Author                                  Contributors  \
0          A. A.                               FORBES, Walter.   
1      A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   
2      A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   
3      A., E. S.                   Appleyard, Ernest Silvanus.   
4      A., E. S.                           BROOME, John Henry.   
5      A., E. S.                           BROOME, John Henry.   
6      A., F. E.                       ASHLEY, Florence Emily.   
7  A., J.|A., J.                 ANDREWS, J. - Writer of Verse   
8         Remaʿ.                                  ADAMS, Mary.   
9         A., T.                                 OLDHAM, John.   

   Corporate Author  Corporate Contributors Former owner  Engraver  \
0               NaN                     NaN          NaN       NaN   
1               NaN                     NaN          NaN       NaN   
2               NaN                     NaN          NaN       NaN   
3               NaN                     NaN          NaN       NaN   
4               NaN                     NaN          NaN       NaN   
5               NaN                     NaN          NaN       NaN   
6               NaN                     NaN          NaN       NaN   
7               NaN                     NaN          NaN       NaN   
8               NaN                     NaN          NaN       NaN   
9               NaN                     NaN          NaN       NaN   

  Issuance type                                         Flickr URL  \
0   monographic  http://www.flickr.com/photos/britishlibrary/ta...   
1   monographic  http://www.flickr.com/photos/britishlibrary/ta...   
2   monographic  http://www.flickr.com/photos/britishlibrary/ta...   
3   monographic  http://www.flickr.com/photos/britishlibrary/ta...   
4   monographic  http://www.flickr.com/photos/britishlibrary/ta...   
5  

When we look at the first five entries using the head() method, we can see that a handful of columns provide ancillary information that would be helpful to the library but isn’t very descriptive of the books themselves: Edition Statement, Corporate Author, Corporate Contributors, Former owner, Engraver, Issuance type and Shelfmarks.

We can drop these columns in the following way:

In [4]:
# create a list to drop columns
to_drop = ['Edition Statement',
            'Corporate Author',
            'Corporate Contributors',
            'Former owner',
            'Engraver',
            'Contributors',
            'Issuance type',          
           'Shelfmarks']
df.drop(to_drop, inplace=True, axis=1)

Above, we defined a list that contains the names of all the columns we want to drop. Next, we call the drop() function on our object, passing in the inplace parameter as True and the axis parameter as 1. This tells Pandas that we want the changes to be made directly in our object and that it should look for the values to be dropped in the columns of the object. 

When we inspect the DataFrame again, we’ll see that the unwanted columns have been removed:

In [5]:
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                         Flickr URL  
0      A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
1  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
2  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
3  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...  
4  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...

Alternatively, we could also remove the columns by passing them to the columns parameter directly instead of separately specifying the labels to be removed and the axis where Pandas should look for the labels:

```python
df.drop(columns=to_drop, inplace=True)
```

#### Changing the Index of a DataFrame

A Pandas Index extends the functionality of NumPy arrays to allow for more versatile slicing and labeling. In many cases, it is helpful to use a uniquely valued identifying field of the data as its index.

For example, in the dataset used in the previous section, it can be expected that when a librarian searches for a record, they may input the unique identifier (values in the Identifier column) for a book:

In [6]:
df['Identifier'].is_unique

True

Let’s replace the existing index with this column using 
```python
pandas.DataFrame.set_index
```

In [ ]:
df = df.set_index('Identifier')
df.head()

**Technical Detail: Unlike primary keys in SQL, a Pandas Index doesn’t make any guarantee of being unique, although many indexing and merging operations will notice a speedup in runtime if it is.**

We can access each record in a straightforward way with loc[]. Although loc[] may not have all that intuitive of a name, it allows us to do label-based indexing, which is the labeling of a row or record without regard to its position:

In [7]:
df.loc[206]

Identifier                                                          77554
Place of Publication                                               London
Date of Publication                                                  1848
Publisher                                                 Richard Bentley
Title                   Rambles in the romantic regions of the Hartz M...
Author                                   Andersen, H. C. (Hans Christian)
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 206, dtype: object

In [ ]:
df.loc?

In other words, 206 is the first label of the index. To access it by position, we could use df.iloc[0], which does position-based indexing.

**Technical Detail: .loc[] is technically a class instance and has some special syntax that doesn’t conform exactly to most plain-vanilla Python instance methods.**

Previously, our index was a RangeIndex: integers starting from 0, analogous to Python’s built-in range. By passing a column name to set_index, we have changed the index to the values in Identifier.

You may have noticed that we reassigned the variable to the object returned by the method with df = df.set_index(...). This is because, by default, the method returns a modified copy of our object and does not make the changes directly to the object. We can avoid this by setting the inplace parameter:

In [9]:
df.set_index('Identifier', inplace=True)

KeyError: 'Identifier'

#### Tidying up Fields in the Data

So far, we have removed unnecessary columns and changed the index of our DataFrame to something more sensible. In this section, we will clean specific columns and get them to a uniform format to get a better understanding of the dataset and enforce consistency. In particular, we will be cleaning Date of Publication and Place of Publication.

Upon inspection, all of the data types are currently the object dtype, which is roughly analogous to str in native Python.

It encapsulates any field that can’t be neatly fit as numerical or categorical data. This makes sense since we’re working with data that is initially a bunch of messy strings:

In [10]:
df.get_dtype_counts()

object    6
dtype: int64

One field where it makes sense to enforce a numeric value is the date of publication so that we can do calculations down the road:

In [11]:
# get the label, column name and return only 10 results
df.loc[1905:, 'Date of Publication'].head(10)

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
2957           1873
3017           1866
3131           1899
4598           1814
4884           1820
Name: Date of Publication, dtype: object

A particular book can have only one date of publication. Therefore, we need to do the following:

- Remove the extra dates in square brackets, wherever present: 1879 [1878]

- Convert date ranges to their “start date”, wherever present: 1860-63; 1839, 38-54

- Completely remove the dates we are not certain about and replace them with NumPy’s NaN: [1897?]

- Convert the string nan to NumPy’s NaN value
Synthesizing these patterns, we can actually take advantage of a single regular expression to extract the publication year:

In [13]:
import re
regex = r'^(\d{4})'

The regular expression above is meant to find any four digits at the beginning of a string, which suffices for our case. The above is a raw string (meaning that a backslash is no longer an escape character), which is standard practice with regular expressions.

The \d represents any digit, and {4} repeats this rule four times. The ^ character matches the start of a string, and the parentheses denote a capturing group, which signals to Pandas that we want to extract that part of the regex. (We want ^ to avoid cases where [ starts off the string.)

Let’s see what happens when we run this regex across our dataset:

In [14]:
extr = df['Date of Publication'].str.extract(r'^(\d{4})', expand=False)
extr.head()

Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

Not familiar with regex? You can [inspect the expression](https://regex101.com/r/3AJ1Pv/1) above at regex101.com and read more at the Python Regular Expressions [HOWTO](https://docs.python.org/3.6/howto/regex.html).

Technically, this column still has object dtype, but we can easily get its numerical version with pd.to_numeric:

In [18]:
df['Date of Publication'] = pd.to_numeric(extr)
df['Date of Publication'].dtype

dtype('float64')

In [19]:
df['Date of Publication'].isnull().sum() / len(df)

0.11717147339205986

Great! That’s done!

Next time will be 

# Combining str Methods with NumPy to Clean Columns

In [20]:
'''Above, you may have noticed the use of df['Date of Publication'].str. 
This attribute is a way to access speedy string operations in Pandas that largely mimic 
operations on native Python strings or compiled regular expressions, such as .split(), .replace(), 
and .capitalize(). To clean the Place of Publication field, we can combine Pandas str methods 
with NumPy’s np.where function, which is basically a vectorized form of Excel’s IF() macro. It has 
the following syntax:

np.where(condition, then, else)'''

SyntaxError: invalid syntax (<ipython-input-20-f51047754b2b>, line 8)

In [23]:
pub = df['Place of Publication']
london = pub.str.contains('London')
oxford = pub.str.contains('Oxford')
pub.head()

Identifier
206                      London
216    London; Virtue & Yorston
218                      London
472                      London
480                      London
Name: Place of Publication, dtype: object

In [24]:
df['Place of Publication'] = np.where(london, 'London', 
                                      np.where(oxford, 'Oxford', 
                                               pub.str.replace('-', ' ')))

In [25]:
df.head()

Place of Publication  Date of Publication              Publisher  \
Identifier                                                                    
206                      London               1879.0       S. Tinsley & Co.   
216                      London               1868.0           Virtue & Co.   
218                      London               1869.0  Bradbury, Evans & Co.   
472                      London               1851.0          James Darling   
480                      London               1857.0   Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...